Add webui

jogibear9988 opened the pull request.

jogibear9988 commented:
> * [ ]  ❗ [E500] [node_modules](https://github.com/iobroker-community-adapters/ioBroker.webui/tree/master/node_modules) found in repo. Please delete it
it's a visualization adapter, so it's needed
> * [ ]  ❗ [E519] "./dist/backend/main.js" found in [package.json](https://github.com/iobroker-community-adapters/ioBroker.webui/blob/master/package.json), but not found as file
look in the package under npmjs, the file is there
> * [ ]  ❗ [E603] NO "## Changelog" found in [README.md](https://github.com/iobroker-community-adapters/ioBroker.webui/blob/master/README.md)
there is a changelog.md
> * [ ]  ❗ [E812] file test/ found in repository, but not found in .npmignore
there is /test
> * [ ]  ❗ [E813] file src/ found in repository, but not found in .npmignore
there is /src
> * [ ]  👀 [W171] "common.title" is deprecated in [io-package.json](https://github.com/iobroker-community-adapters/ioBroker.webui/blob/master/io-package.json)
yeah, but every other packe still also contains it
> * [ ]  👀 [W109] "common.desc" should be translated into all supported languages (en, de, ru, pt, nl, fr, it, es, pl, uk, zh-cn)
thought the release script will do this? i will not do manually
> * [ ]  👀 [W400] Cannot find "webui" in latest repository
??? don't understand this error

mcm1957 commented:
To answer questions asked at telegram:

Every new adapter will undergo an automatic revoew by adapter vhecker and a?manual review. Manual revie will last some time. Please expect at least one or two weeks for a first review.

To minimize issues during review read the Readme of this repository carefully and check section requiremts for an adapter to be added to latest.

AlCalzone commented:
> it's a visualization adapter, so it's needed

Why? node_modules includes unnecessary stuff.

> there is a changelog.md

Some parts of the infrastructure/admin expect the most recent part of the changelog to be in the README. You can check how I solved in https://github.com/AlCalzone/ioBroker.zwave2/ - The last few entries are in the README, the rest automatically gets moved to CHANGELOG_old.md by the release script.

> ...npmignore

I recommend using the `files` field in `package.json` instead. Saying which files you want in the npm module instead of hoping you ignored all the ones you don't is less error-prone. Also the checker shouldn't complain then.

> thought the release script will do this? i will not do manually

No. https://github.com/ioBroker/adapter-dev/ does, or just put the section in https://translator-ui.iobroker.in/ - only takes 2 minutes.

jogibear9988 commented:
> > it's a visualization adapter, so it's needed
> 
> Why? node_modules includes unnecessary stuff. If you need those, look into a bundler, e.g. [esbuild](https://esbuild.github.io/)

no, my visu uses importmaps and no bundler. this is a architactural design decission

jogibear9988 commented:
RE-CHECK!

jogibear9988 commented:
RE-CHECK!

mcm1957 commented:
E500 will be part of review by Apillon77
W400 can be ignored of course for flatest PR

jogibear9988 commented:
To E500

At the moment the inclusion of the node_modules folder could not be changed, as I'm not using a Bundler. Also it is not possible at the Moment to bundle the project, as how the loading of some of the modules happen in the webcomponent designer.
These may change in the feature when import attributes are supported by the platform and the bundlers. But also if someone uses HTTP2 the loading should be not so much slower, so I don't know if it is worth a bundling step.

And also, you could later add npm-packages to the designer, wich then also could depend on packages wich are included in the designer, this will not be possible when the designer is bundeld.

jogibear9988 commented:
@Apollon77 Is it okay now with E500? I can nothing do about this...

mcm1957 commented:
Please wait until you get review feedback.
Your adapter is not the only one waitung. 

Apollon77 will do a review as soon as he has time and report all things he thinks are important to fix.

mcm1957 commented:
RE-CHECK!

mcm1957 commented:
First of all - THANK YOU for the time and effort you spend to maintain this adapter.

I would like to give some (preliminary) feedback based on my personal oppinion. This is NOT an offical review and @Apollon77 might have several additional suggestions or even a different oppinion to one or the other statement. In addtion this revie ONLY refers to the backend / adapter part - the frontend / www part must be reviewed by @Apollon77.

Please feel free to contact him (or wait for a response from him) if you cannot follow my suggestsions or before you spend major effort.

- [ ] errors reported by adapter checker must be fixed / evaluated

  ```[E500] node_modules found in repo. Please delete it``` refers to  https://github.com/iobroker-community-adapters/ioBroker.webui/tree/master/www/node_modules. node_modules inside adapter direvctory are invalid. Whether a node_moduls directory insode www tree is acceptable must be evaluated / decided by @Apollon77.

  So please wait for his statement regarding this point.
  
- [ ] adapter has 'singleton' set to true / check compact attribute

  As this adapter is allowed to have only one instance, creates subprocesses and seems to need significant resources please evaluate whether its really useful to enable compact mode for this adapter. Eventually set compact = false unless you see real benefits and have tested the adapter in compact mode.

- [ ] reevaluate read/write attribute of states

  All states defined at io-package.json have write attribute set to true. Its unlikely, that all those states are designed to operate as input for the adapter and should be written ba other adapters, i.e. state.npm is a feedback state used during npm installation - if I read the code correctly. (https://github.com/iobroker-community-adapters/ioBroker.webui/blob/b49e969ebb297e288a0316a506bd4d95da12ed51/io-package.json#L251)

- [ ] onUnload code needs rewrite

https://github.com/iobroker-community-adapters/ioBroker.webui/blob/b49e969ebb297e288a0316a506bd4d95da12ed51/src/backend/main.ts#L201C1-L204C1

  Rountine onUnload code must stop all running timers (seem none exist) and spawned subprocessed (i.e. npm installation). In addition onUnload must call the provide callback when completed. See templates or originally created adapter code for details.

  In addtion evaluate the use of variable _onUnload. It is written once, but I did not find any usage.

- [ ] adapt testing to supported node releases

  Tests for node 16 is mandatory unless you require node 18 and higher (at package.json).
  Tests for node 18 is mandatory as this is the recommended node version and node 16 will be eol and deprected Q3/2023.
  Consider adding node 20 tests unless you already know incompatibilities.
  
  So the recommended testmatrix is [16.x, 18.x, 20.x] or [18.x, 20.x] depending on engines requirments setting at package.json 

  As node 16 is end-of-life now, it's ok to remove node 16 from testing. In this case please require node 18 as minimum node version at package.json. Currently node 16 is required / allowed) but not tested.
  
In general I must state once more, that this review ONLY refers to the adapter / backend part. The frontend / www part need to be reviewed by @Apollon77.

Thanks for reading and evaluating this suggestions.
McM1957

**Please add a comment when you have reviewed and fixed the suggestionsor at least commented the suggestions and you think the adapter is ready for a re-review!**


**Additional notess addressed at @Apollon77 - please consider them during your review** 

- [ ] delivering node_modules below www

  Please check whether delivering node_modules below www directory tree is ok or not and eventually adapterchecker should be adapted to handle this (if delivery is ok)

- [ ] using state as command interface

  Is using several states as command interface ok? Or wouldn't this be handled by message in a better way?

- [ ] on the fly installation (running npm from adapter)

  Is it ok that an adapter installs npm packages while the system is running? 
  Are there sufficient firewalls that the due to some error or user misuse the adapter cannot open scurity holes (i.e. installing anything without sufficient access rights of the user at the system) ?
  Are there sufficient firewalls that npm installation does not interfere with oterh adapters or ioBroker core ?

- [ ] complete frontend (www) including installation of npm and uploading needs review. I cannot check this due to missing knowhow in this area. Apollon77 please take this part.

jogibear9988 commented:
> 



> First of all - THANK YOU for the time and effort you spend to maintain this adapter.
> 
> I would like to give some (preliminary) feedback based on my personal oppinion. This is NOT an offical review and @Apollon77 might have several additional suggestions or even a different oppinion to one or the other statement. In addtion this revie ONLY refers to the backend / adapter part - the frontend / www part must be reviewed by @Apollon77.
> 
> Please feel free to contact him (or wait for a response from him) if you cannot follow my suggestsions or before you spend major effort.
> 
> * [ ]  errors reported by adapter checker must be fixed / evaluated
>   `[E500] node_modules found in repo. Please delete it` refers to  https://github.com/iobroker-community-adapters/ioBroker.webui/tree/master/www/node_modules. node_modules inside adapter direvctory are invalid. Whether a node_moduls directory insode www tree is acceptable must be evaluated / decided by @Apollon77.
>   So please wait for his statement regarding this point.
> * [ ]  adapter has 'singleton' set to true / check compact attribute
>   As this adapter is allowed to have only one instance, creates subprocesses and seems to need significant resources please evaluate whether its really useful to enable compact mode for this adapter. Eventually set compact = false unless you see real benefits and have tested the adapter in compact mode.

-> I have tested it, and most of the time the adapter does nothing. Only when you add or remove a NPM package


> * [ ]  reevaluate read/write attribute of states
>   All states defined at io-package.json have write attribute set to true. Its unlikely, that all those states are designed to operate as input for the adapter and should be written ba other adapters, i.e. state.npm is a feedback state used during npm installation - if I read the code correctly. (https://github.com/iobroker-community-adapters/ioBroker.webui/blob/b49e969ebb297e288a0316a506bd4d95da12ed51/io-package.json#L251)

-> not all, some of them (the onse for the files have none, it's only the npm state who is wrong, I fixed it)

> * [ ]  onUnload code needs rewrite
> 
> https://github.com/iobroker-community-adapters/ioBroker.webui/blob/b49e969ebb297e288a0316a506bd4d95da12ed51/src/backend/main.ts#L201C1-L204C1
> 
> Rountine onUnload code must stop all running timers (seem none exist) and spawned subprocessed (i.e. npm installation). In addition onUnload must call the provide callback when completed. See templates or originally created adapter code for details.
> 
> In addtion evaluate the use of variable _onUnload. It is written once, but I did not find any usage.

-> I look that I cancel the processes

> 
> * [ ]  adapt testing to supported node releases
>   Tests for node 16 is mandatory unless you require node 18 and higher (at package.json).
>   Tests for node 18 is mandatory as this is the recommended node version and node 16 will be eol and deprected Q3/2023.
>   Consider adding node 20 tests unless you already know incompatibilities.
>   So the recommended testmatrix is [16.x, 18.x, 20.x] or [18.x, 20.x] depending on engines requirments setting at package.json
>   As node 16 is end-of-life now, it's ok to remove node 16 from testing. In this case please require node 18 as minimum node version at package.json. Currently node 16 is required / allowed) but not tested.
> 
> In general I must state once more, that this review ONLY refers to the adapter / backend part. The frontend / www part need to be reviewed by @Apollon77.
> 
> Thanks for reading and evaluating this suggestions. McM1957
> 
> **Please add a comment when you have reviewed and fixed the suggestionsor at least commented the suggestions and you think the adapter is ready for a re-review!**
> 
> **Additional notess addressed at @Apollon77 - please consider them during your review**
> 

-> I will not add test for 16

> * [ ]  delivering node_modules below www
>   Please check whether delivering node_modules below www directory tree is ok or not and eventually adapterchecker should be adapted to handle this (if delivery is ok)

-> I wll not look into adapterchecker

> * [ ]  using state as command interface
>   Is using several states as command interface ok? Or wouldn't this be handled by message in a better way?

-> I would it have similar to vis, so I leave it like this

> * [ ]  on the fly installation (running npm from adapter)
>   Is it ok that an adapter installs npm packages while the system is running?
>   Are there sufficient firewalls that the due to some error or user misuse the adapter cannot open scurity holes (i.e. installing anything without sufficient access rights of the user at the system) ?
>   Are there sufficient firewalls that npm installation does not interfere with other adapters or ioBroker core ?

-> I could use Security of iobroker (if there is one), but yes it is needed like it is in the adapter. 

> * [ ]  complete frontend (www) including installation of npm and uploading needs review. I cannot check this due to missing knowhow in this area. Apollon77 please take this part.

mcm1957 commented:
merged on error
Will create a new PR ...

mcm1957 commented:
PR https://github.com/ioBroker/ioBroker.repositories/pull/2589 created as replacement

jogibear9988 commented:
To security once more:

The NPM Pakets wich are installed, are only searched and parsed, not run on the server. Only uploaded and then the code is run on the client. So I think not really a security concern.
Also the webui backend could be disabled, and only run when needed. It's only used for the NPM installation. I could think about that the client itself downloads the NPM and uploads to iobroker files, so not backend is needed at all. But this will not be done in next time.

mcm1957 commented:
please use PR https://github.com/ioBroker/ioBroker.repositories/pull/2589 created as replacement for further discussions.
